### PR TITLE
feat(unstable): `--unstable-detect-cjs` for respecting explicit `"type": "commonjs"`

### PR DESCRIPTION
**Pull request by dsherret**
When using the `--unstable-detect-cjs` flag or adding `"unstable": ["detect-cjs"]` to a deno.json, it will make a JS file CJS if the closest package.json contains `"type": "commonjs"` and the file is not an ESM module (no TLA, no `import.meta`, no `import`/`export`).

* https://github.com/denoland/deno/issues/26225

### CODE REVIEW

File: cli/cache/mod.rs [hunk 1]
```diff
@@ -282,6 +291,52 @@ impl Loader for FetchCacher {
...
+                      Keyword::Export | Keyword::Import
+                    ))
+                  )
+                });
+                if !is_es_module {
+                  return Box::pin(futures::future::ready(Ok(Some(
+                    LoadResponse::External {
+                      specifier: specifier.clone(),
+                    },
+                  ))));
```
Comment (dsherret) [inline cli/cache/mod.rs:333]: Actually, this probably can all be done in the deno_core loader.

File: cli/args/flags.rs [hunk 2]
```diff
@@ -4192,8 +4207,8 @@ impl Iterator for UnstableArgsIter {
         UnstableArgsConfig::ResolutionOnly | UnstableArgsConfig::ResolutionAndRuntime => Some("true")
       })
       .help_heading(UNSTABLE_HEADING)
-    } else if self.idx > 3 {
-      let granular_flag = crate::UNSTABLE_GRANULAR_FLAGS.get(self.idx - 4)?;
+    } else if self.idx > 4 {
+      let granular_flag = crate::UNSTABLE_GRANULAR_FLAGS.get(self.idx - 5)?;
```
Comment (dsherret) [inline cli/args/flags.rs:4211]: Would be better if the code didn't require changing numbers like this.

Comment (bartlomieju) [reply cli/args/flags.rs:4211]: CC @crowlKats

**Review by marvinhagemeister — COMMENTED** [review]
Gave this a go on a test project and it works great! The config schema in `cli/schemas/config-file.v1.json` might also need to be update to get that fancy autocompletion for the unstable features.

File: cli/args/flags.rs (hunk 2 above)
Comment (crowlKats) [reply cli/args/flags.rs:4211]: this could just be an `else` case instead

Comment (dsherret) [reply cli/args/flags.rs:4211]: There is already an `else`. Why do we do this instead of just constructing a vector?

Comment (dsherret) [reply cli/args/flags.rs]: I removed this code in https://github.com/denoland/deno/pull/26149/commits/fdbb7a42a5658536b1dcf08af32e69a2c9e07910

File: cli/args/flags.rs [hunk 3]
```diff
@@ -4142,85 +4143,107 @@ enum UnstableArgsConfig {
...
+          UnstableArgsConfig::ResolutionOnly
+          | UnstableArgsConfig::ResolutionAndRuntime => Some("true"),
+        })
+        .help_heading(UNSTABLE_HEADING)
+        .display_order(next_display_order()),
+    ).arg(
+      Arg::new("unstable-detect-cjs")
+        .long("unstable-detect-cjs")
+        .help("Reads the package.json type field in a project to treat .js files as .cjs")
+        .env("DENO_UNSTABLE_PACKAGE_JSON_TYPE")
```
Comment (bartlomieju) [inline cli/args/flags.rs:4187]: Do we really need it? Is it used only for testing? If so, maybe we should handle it manually so it doesn't show up in `--help` output?

**Review by bartlomieju — COMMENTED** [review]
Overall looks good - could you please add some test cases for `deno compile`? It appears there were changes in that code as well that are currently not exercised.

File: cli/args/flags.rs (hunk 3 above)
Comment (dsherret) [reply cli/args/flags.rs:4187]: Oops, this should be the new name, but also I don't think we should bother with these env vars. It probably hurts startup time and I bet nobody uses it.

**Comment by dsherret** [comment]
> Overall looks good - could you please add some test cases for deno compile? It appears there were changes in that code as well that are currently not exercised.

I'll add a message to say it's not supported in deno compile (right now it will crash).

File: cli/args/flags.rs (hunk 3 above)
Comment (bartlomieju) [reply cli/args/flags.rs:4187]: I agree, let's remove it then.

**Comment by bartlomieju** [comment]
> > Overall looks good - could you please add some test cases for deno compile? It appears there were changes in that code as well that are currently not exercised.
> 
> I'll add a message to say it's not supported in deno compile (right now it will crash in most cases).

Sounds good.

**Pull request merged**